### PR TITLE
feat: Link forum profile with DAO profile

### LINKED ISSUE #926: Identity: Link discourse user to dApp user
First slice of the identity overhaul project:

- Manual flow prompted by the app and executed by the user 
- The flow needs to prove that the owner of the logged in wallet in the Governance dApp also controls the Discourse account
- Once the flow is completed and the accounts are linked, user should be able to post a comment (as it is right now, no changes) and other users who read the comment should be able to click on their username (TBD: Discord or DCL username) and go to the Governance dApp profile page.

>

New model user `{ address, discourse_user }`
Settings page where I can link my discourse account.
Link discourse button will take user to discourse authorization page with a public key as query param. See implementation in https://github.com/KengoTODA/discourse-api-key-generator/blob/main/src/index.ts. 
We need to set up an auth redirect URL where we can read the auth payload to verify the user (possible /settings/integrations, /settings/discourse).
Fetch to endpoint in governance backend with the payload, decrypt payload with keys, use `https://forum.decentraland.vote/session/current.json` with `User-Api-Key` as header to get username. Store username in user row.

### PR DESCRIPTION
**Pull request by ncomerci**
This closes #926

### CODE REVIEW

File: src/entities/Discourse/routes.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import { ValidationMessage } from './types'
+
+const TIMEOUT_TIME = 5 * 60 * 1000 // 5mins
+const VALIDATIONS_IN_PROGRESS: Record<string, ValidationMessage> = {}
+
+export default routes((route) => {
+  const withAuth = auth()
+  route.get('/forumId', withAuth, handleAPI(getForumId))
+  route.get('/validateProfile', withAuth, handleAPI(getValidationMessage))
+  route.post('/validateProfile', withAuth, handleAPI(checkValidationMessage))
```
Comment (andyesp) [inline src/entities/Discourse/routes.ts:18]: We are using kebab-case for backend routes.

File: src/entities/Discourse/model.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import { Model } from 'decentraland-gatsby/dist/entities/Database/model'
+import { SQL, join, table } from 'decentraland-gatsby/dist/entities/Database/utils'
+
+import { DiscourseAttributes } from './types'
+
+export default class DiscourseModel extends Model<DiscourseAttributes> {
```
Comment (andyesp) [inline src/entities/Discourse/model.ts:6]: How about changing this to UserModel so we can expand it later with the other connections?

File: src/entities/Discourse/types.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+export type DiscourseAttributes = {
+  address: string
+  forum_id: number
```
Comment (andyesp) [inline src/entities/Discourse/types.ts:3]: Is this `forum_username` instead? 🤔

File: src/clients/Governance.ts [hunk 4]
```diff
@@ -85,6 +86,15 @@ export class Governance extends API {
     return this.Cache.get(baseUrl)!
   }
 
+  // private async trycatch<T>(fn: () => Promise<T>, fallback: T): Promise<T> {
```
Comment (andyesp) [inline src/clients/Governance.ts:89]: 🫣

File: src/entities/Discourse/types.ts (hunk 3 above)
Comment (ncomerci) [reply src/entities/Discourse/types.ts:3]: `forum_username` could eventually change, forum_id is immutable

File: src/intl/en.json [hunk 5]
```diff
@@ -561,6 +565,59 @@
...
+        "success_text": "**Governance and Forum accounts linked successfully** 🎉\n\nNow your comments left on Proposals' forum threads will link back to, and be displayed on your Governance profile.",
+        "success_subtext": "Plus, you'll get a fancy icon next to your **username**",
+        "success_button": "Got it",
+        "error_text": "**Governance and Forum accounts could not be linked**\n\nSomething happened. It must have been aliens, or a malevolent AI but we can't be sure.",
+        "error_subtext": "Maybe just have a go at it one more time?",
+        "error_button": "Retry"
+      },
+      "discord": {
+        "card_title": "Discord",
+        "card_description": "Elevate your credibility by letting others know which wallet is behind your chat interactions."
```
Comment (andyesp) [inline src/intl/en.json:614]: Maybe `address` instead of wallet?

Comment (andyesp) [reply src/intl/en.json:614]: We also use the `name` when there is one available.

File: src/entities/Proposal/routes.ts [hunk 6]
```diff
@@ -531,17 +533,36 @@ export async function removeProposal(req: WithAuth<Request<{ proposal: string }>
...
+
+    const userIds = new Set(filteredComments.comments.map((comment) => comment.id))
+
+    const users = await DiscourseModel.getAddressesByForumId(Array.from(userIds))
+    const userMapping = users.reduce((map, user) => {
+      map[user.forum_id] = user.address
+      return map
+    }, {} as Record<number, string>)
+
+    return {
```
Comment (andyesp) [inline src/entities/Proposal/routes.ts:550]: Shouldn't this return `totalComments` too?

File: src/entities/Proposal/routes.ts [hunk 7]
```diff
@@ -531,17 +533,36 @@ export async function removeProposal(req: WithAuth<Request<{ proposal: string }>
...
+      ...filteredComments,
+      comments: filteredComments.comments.map((comment) => {
```
Comment (andyesp) [inline src/entities/Proposal/routes.ts:552]: Could this be part of `filterComments`?

File: src/clients/Governance.ts [hunk 8]
```diff
@@ -437,6 +437,42 @@ export class Governance extends API {
...
+  async isProfileValidated(address: string) {
+    const result = await this.fetch<ApiResponse<boolean>>(
+      `/is-profile-validated/${address}`,
+      this.options().method('GET')
+    )
+
+    return result.data
+  }
+
+  async removeVerification(address: string) {
```
Comment (andyesp) [inline src/clients/Governance.ts:467]: We should remove this after prod deploy right?

File: src/components/User/Username.css [hunk 9]
```diff
@@ -24,3 +24,7 @@ a.Username span.address {
 .Username .dcl.blockie-wrapper .dcl.blockie-children {
   margin-left: 0.5rem;
 }
+
+.Username svg {
```
Comment (andyesp) [inline src/components/User/Username.css:28]: Is this modifyng the Avatar? Can you pass a className to Avatar instead?

File: src/entities/User/routes.ts [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+import { DiscourseService } from '../../services/DiscourseService'
+import isDebugAddress from '../Debug/isDebugAddress'
+
+import { GATSBY_DISCOURSE_CONNECT_THREAD, MESSAGE_TIMEOUT_TIME } from './constants'
+import UserModel from './model'
+import { ValidationMessage } from './types'
+
+export default routes((route) => {
+  const withAuth = auth()
+  route.get('/is-profile-validated/:address', handleAPI(isProfileValidated))
```
Comment (andyesp) [inline src/entities/User/routes.ts:17]: What if this is a `/profile` endpoint instead that we could check if a property exists or has a certain value to see if is validated? We can call this in a `useProfile` hook.

File: src/components/Modal/ExternalLinkWarningModal/ExternalLinkWarningModal.tsx [hunk 11]
```diff
@@ -37,7 +35,7 @@ function ExternalLinkWarningModal() {
   const { isWarningModalOpen, href } = warningModalState
   const handleDismiss = () => setWarningModalState((state) => ({ ...state, isWarningModalOpen: false }))
   const handleContinue = useCallback(() => {
-    openInNewTab(href)
+    openUrl(href)
```
Comment (1emu) [inline src/components/Modal/ExternalLinkWarningModal/ExternalLinkWarningModal.tsx:38]: I think this should be `openUrl(href, true)`

File: src/pages/debug.tsx [hunk 12]
```diff
@@ -34,6 +35,7 @@ export default function DebugPage() {
       <EnvStatus className="DebugPage__Section" />
       <BudgetsUpdate className="DebugPage__Section" />
       <ErrorReporting className="DebugPage__Section" />
+      <RemoveVerification className="DebugPage__Section" />
```
Comment (1emu) [inline src/pages/debug.tsx:38]: This should also be remove before prod right?

File: src/intl/en.json (hunk 5 above)
Comment (1emu) [reply src/intl/en.json:614]: How about `account`? (ethereum addresses are also known as externally owned accounts, so it works on both levels)

File: src/hooks/useForumConnect.ts [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+  DISCOURSE_API.endsWith('/') ? '' : '/'
+}t/${GATSBY_DISCOURSE_CONNECT_THREAD}/10000`
+const VALIDATION_CHECK_INTERVAL = 5 * 1000 // 5 seconds
+
+const getMessage = async () => Governance.get().getValidationMessage()
+export default function useForumConnect() {
+  const [user, userState] = useAuthContext()
+  const [sign, signState] = useSign(user, userState.provider)
+  const [copied, clipboardState] = useClipboardCopy(Time.Second)
+  const [signResolveReject, setSignResolveReject] = useState<{
```
Comment (1emu) [inline src/hooks/useForumConnect.ts:25]: How about giving this resolve/reject object a name? (and the state/setState too)
`SignatureResolution` makes sense?

**Review by 1emu — APPROVED** [review]
Nicee Nicoo! Left some comments/suggestions :)

File: src/components/Icon/Identity.tsx [hunk 14]
```diff
@@ -0,0 +1,103 @@
+/* eslint-disable react/no-unknown-property */
+import React from 'react'
+
+function Identity() {
+  return (
+    <svg width="418" height="185" viewBox="0 0 418 185" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (andyesp) [inline src/components/Icon/Identity.tsx:6]: Is this svg this image?
image
It's not an icon 🤔

File: src/components/Modal/IdentityConnectModal/AccountsConnectModal.tsx [hunk 15]
```diff
@@ -0,0 +1,350 @@
...
+    action: actions[index],
+    isDisabled: index + 1 > currentStep,
+    isCompleted: index + 1 < currentStep,
+    helper: stepHelpers?.[index + 1],
+  }))
+}
+
+function getHelperTexts(account: AccountType, stepsAmount: number): string[] {
+  if (stepsAmount === 0) {
+    console.error('Invalid steps amount')
```
[the remark this answers is not in the document]
Comment (ncomerci) [reply src/components/Modal/IdentityConnectModal/AccountsConnectModal.tsx:115]: It is an error that helps during development but in prod it should not fail. I think reporting to rollbar is unnecessary

image

File: src/components/Modal/IdentityConnectModal/AccountsConnectModal.tsx [hunk 16]
```diff
@@ -0,0 +1,350 @@
...
+    try {
+      setIsTimerActive(true)
+      setStepHelper(STEP_NUMBER, 'active')
+      await getSignedMessage()
+      setStepHelper(STEP_NUMBER, 'success')
+      nextStep()
+    } catch (error) {
+      setIsTimerActive(false)
+      setStepHelper(STEP_NUMBER, 'error')
+      console.error(error)
```
[the remark this answers is not in the document]
Comment (ncomerci) [reply src/components/Modal/IdentityConnectModal/AccountsConnectModal.tsx:198]: This error is usually when the user refuses to sign the message, rarely for any other reason. Including rollbar here would mean a notification every time someone refuses to sign the message 🤔

File: src/components/Modal/IdentityConnectModal/AccountsConnectModal.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (andyesp) [reply src/components/Modal/IdentityConnectModal/AccountsConnectModal.tsx:115]: I'd remove the console.error.

File: src/components/Modal/ExternalLinkWarningModal/ExternalLinkWarningModal.tsx (hunk 11 above)
Comment (ncomerci) [reply src/components/Modal/ExternalLinkWarningModal/ExternalLinkWarningModal.tsx:38]: By default it is true!

image

File: src/pages/debug.tsx (hunk 12 above)
Comment (ncomerci) [reply src/pages/debug.tsx:38]: Yep!

File: src/components/Modal/IdentityConnectModal/AccountsConnectModal.tsx [hunk 17]
```diff
@@ -0,0 +1,370 @@
...
+  return Array.from({ length: stepsAmount }, (_, index) => `modal.identity_setup.${account}.helper_step_${index + 1}`)
+}
+
+function getTimeFormatted(totalSeconds: number) {
+  const minutes = Math.floor(totalSeconds / 60)
+  const seconds = totalSeconds % 60
+  return `${minutes}:${seconds < 10 ? `0${seconds}` : seconds}`
+}
+
+function AccountsConnectModal({ open, onClose }: ModalProps & { onClose: () => void }) {
```
Comment (andyesp) [inline src/components/Modal/IdentityConnectModal/AccountsConnectModal.tsx:152]: Isn't `onClose` part of `ModalProps`?

File: src/components/Modal/IdentityConnectModal/AccountsConnectModal.tsx [hunk 18]
```diff
@@ -0,0 +1,370 @@
...
+    isValidated,
+    reset: resetForumConnect,
+  } = useForumConnect()
+
+  const [modalState, setModalState] = useState<ModalState>(INITIAL_STATE)
+
+  const setCurrentStep = (currentStep: number) => setModalState((state) => ({ ...state, currentStep }))
+  const setIsTimerActive = (isTimerActive: boolean) => setModalState((state) => ({ ...state, isTimerActive }))
+  const setCurrentType = (currentType: ModalType) => setModalState((state) => ({ ...state, currentType }))
+  const setIsValidating = (isValidating: boolean) => setModalState((state) => ({ ...state, isValidating }))
```
Comment (andyesp) [inline src/components/Modal/IdentityConnectModal/AccountsConnectModal.tsx:170]: I wonder if by doing this is the same as separating the states...
```
const [currentStep, setCurrentStep] = useState()
...etc.
```

File: src/components/Modal/IdentityConnectModal/AccountsConnectModal.tsx [hunk 19]
```diff
@@ -0,0 +1,370 @@
...
+    setModalState((state) => ({ ...state, stepsCurrentHelper: { ...state.stepsCurrentHelper, ...newHelpers } }))
+  }
+  const resetState = (account: AccountType) => {
+    setModalState((state) => ({ ...state, currentStep: 1 }))
+    setIsValidating(false)
+    initializeStepHelpers(account)
+  }
+  const isTimerExpired = time <= 0
+
+  const timerTextKey = useMemo(
```
Comment (andyesp) [inline src/components/Modal/IdentityConnectModal/AccountsConnectModal.tsx:195]: I don't think this memo is necessary.

File: src/entities/User/routes.ts [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+      throw new Error('Validation timed out')
+    }
+
+    const { address, timestamp } = messageProperties
+
+    const comments = await DiscourseService.fetchAllComments(Number(GATSBY_DISCOURSE_CONNECT_THREAD))
+    const timeWindow = new Date(new Date().getTime() - MESSAGE_TIMEOUT_TIME)
+
+    const filteredComments = comments.filter((comment) => new Date(comment.created_at) > timeWindow)
+    const validComment = filteredComments.find((comment) => {
```
Comment (1emu) [inline src/entities/User/routes.ts:71]: You could extract this bit to another function:

```typescript
 const timeWindow = new Date(new Date().getTime() - MESSAGE_TIMEOUT_TIME)

    const filteredComments = comments.filter((comment) => new Date(comment.created_at) > timeWindow)
    const validComment = filteredComments.find((comment) => {
      const addressRegex = new RegExp(address, 'i')
      const dateRegex = new RegExp(timestamp, 'i')

      return addressRegex.test(comment.cooked) && dateRegex.test(comment.cooked)
    })
```
    
Something like

```typescript
const validationComment = getValidationComment(comments, address, timestamp);
```

We can later put this in a UserService... 👀

File: src/components/Modal/IdentityConnectModal/AccountsConnectModal.tsx (hunk 18 above)
Comment (ncomerci) [reply src/components/Modal/IdentityConnectModal/AccountsConnectModal.tsx:170]: Yeah, should be the same but I think is clearer to have a defined type "state"

```typescript
type ModalState = {
  currentType: ModalType
  currentStep: number
  isTimerActive: boolean
  isValidating: boolean
  stepsCurrentHelper?: Record<number, string>
}
```

File: src/entities/User/routes.ts [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+      const recoveredAddress = recoverAddress(hashMessage(formatValidationMessage(address, timestamp)), signature)
+
+      if (recoveredAddress.toLowerCase() !== user) {
+        throw new Error('Invalid signature')
+      }
+
+      await UserModel.createConnection(user, validComment.user_id)
+      clearValidationInProgress(user)
+    }
+
```
Comment (1emu) [inline src/entities/User/routes.ts:90]: The validation part could also be extracted 

```typescript
const comments = await DiscourseService.fetchAllComments(Number(GATSBY_DISCOURSE_CONNECT_THREAD))
    const validationComment = getValidationComment(comments, address, timestamp)
    await validate(validationComment, address, timestamp, user)

    return {
      valid: !!validationComment,
    }
```

And use `isSameAddress` for this part

```typescript
if (!isSameAddress(recoveredAddress, user)) {
      throw new Error('Invalid signature')
    }
```

File: src/entities/User/types.ts [hunk 22]
```diff
@@ -0,0 +1,16 @@
+export type UserAttributes = {
+  address: string
+  forum_id: number
+  verification_date: string
+}
+
+export type ValidationMessage = {
+  address: string
+  timestamp: string
+  messageTimeout: NodeJS.Timeout
```
Comment (1emu) [inline src/entities/User/types.ts:10]: It's not entirely clear to me which case we should use here.. I think `message_timeout`

File: src/entities/Proposal/types.ts [hunk 23]
```diff
@@ -753,6 +753,8 @@ export const newProposalLinkedWearablesScheme = {
 }
 
 export type ProposalComment = {
+  userForumId: number
```
Comment (1emu) [inline src/entities/Proposal/types.ts:756]: `user_forum_id`

File: src/components/Modal/IdentityConnectModal/AccountsConnectModal.tsx (hunk 17 above)
Comment (ncomerci) [reply src/components/Modal/IdentityConnectModal/AccountsConnectModal.tsx:152]: Yes, but it must receive an event and data, which does not allow me to call the function simply as "onClose()", that's why I redefined it

image

File: src/components/Layout/Navigation.tsx [hunk 24]
```diff
@@ -46,7 +77,32 @@ const Navigation = ({ activeTab }: NavigationProps) => {
           </Link>
           {user && (
             <Link href={locations.profile({ address: user })}>
-              <Tabs.Tab active={activeTab === NavigationTab.Profile}>{t('navigation.profile')}</Tabs.Tab>
+              <Popup
+                style={{ zIndex: 1000 }}
```
Comment (andyesp) [inline src/components/Layout/Navigation.tsx:81]: Can you move this to the `NavigationProfilePopUp` class in the css file?

File: src/components/Modal/IdentityConnectModal/AccountsConnectModal.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (andyesp) [reply src/components/Modal/IdentityConnectModal/AccountsConnectModal.tsx:225]: I'd remove the console.error.

File: src/components/Modal/IdentityConnectModal/AccountsConnectModal.tsx [hunk 25]
```diff
@@ -0,0 +1,370 @@
...
+    if (isValidated) {
+      openUrl(locations.profile({ address: address || '' }), false)
+    } else {
+      resetForumConnect()
+      setModalState(INITIAL_STATE)
+    }
+  }
+
+  return (
+    <>
```
Comment (andyesp) [inline src/components/Modal/IdentityConnectModal/AccountsConnectModal.tsx:300]: Fragment not necessary

File: src/entities/User/model.ts [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+import { SQL, join, table } from 'decentraland-gatsby/dist/entities/Database/utils'
+
+import { UserAttributes, ValidatedAccount } from './types'
+
+export default class UserModel extends Model<UserAttributes> {
+  static tableName = 'users'
+  static withTimestamps = false
+  static primaryKey = 'address'
+
+  static async createConnection(address: string, forum_id: number) {
```
Comment (andyesp) [inline src/entities/User/model.ts:11]: `createForumConnection`?

File: src/hooks/useIsProfileValidated.ts [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+
+import { Governance } from '../clients/Governance'
+
+function useIsProfileValidated(address: string | null) {
+  const [isLoading, setIsLoading] = useState(true)
+  const [isProfileValidated] = useAsyncMemo(
+    async () => {
+      setIsLoading(true)
+      let isValidated = false
+      if (address) {
```
Comment (andyesp) [inline src/hooks/useIsProfileValidated.ts:13]: Can't we assume address exists here if you are calling this with truthy deps and the address as a dep?

File: src/entities/User/routes.ts [hunk 28]
```diff
@@ -0,0 +1,125 @@
...
+    delete VALIDATIONS_IN_PROGRESS[user]
+  }
+}
+
+function formatValidationMessage(address: string, timestamp: string) {
+  return `By signing and posting this message I'm linking my Decentraland DAO account ${address} with this Discourse forum account\n\nDate: ${timestamp}`
+}
+
+async function getValidationMessage(req: WithAuth) {
+  const address = req.auth!
```
Comment (andyesp) [inline src/entities/User/routes.ts:39]: We should check if there is an address here.

File: src/entities/User/routes.ts [hunk 29]
```diff
@@ -0,0 +1,125 @@
...
+  VALIDATIONS_IN_PROGRESS[address] = {
+    ...message,
+    messageTimeout,
+  }
+
+  return formatValidationMessage(address, message.timestamp)
+}
+
+async function checkValidationMessage(req: WithAuth) {
+  const user = req.auth!
```
Comment (andyesp) [inline src/entities/User/routes.ts:58]: Should we check if user exists?

File: src/entities/User/routes.ts (hunk 21 above)
Comment (andyesp) [reply src/entities/User/routes.ts:90]: Agree with the `getValidationComment` function but not sure with the `validate` function.

File: src/entities/User/routes.ts [hunk 30]
```diff
@@ -0,0 +1,125 @@
...
+
+async function isValidated(req: Request) {
+  const address = req.params.address
+  if (!isEthereumAddress(address)) {
+    return false
+  }
+  try {
+    return await UserModel.isProfileValidated(address)
+  } catch (error) {
+    ErrorService.report('Error while fetching validation data', error)
```
Comment (andyesp) [inline src/entities/User/routes.ts:107]: Shouldn't we throw an error here?

File: src/entities/User/routes.ts [hunk 31]
```diff
@@ -0,0 +1,125 @@
...
+    }
+  } catch (error) {
+    throw new Error("Couldn't validate the user. " + error)
+  }
+}
+
+async function isValidated(req: Request) {
+  const address = req.params.address
+  if (!isEthereumAddress(address)) {
+    return false
```
Comment (andyesp) [inline src/entities/User/routes.ts:102]: We should throw an error here

File: src/hooks/useForumConnect.ts [hunk 32]
```diff
@@ -0,0 +1,130 @@
...
+  const [signatureResolution, setSignatureResolution] = useState<{
+    resolve: (value: unknown) => void
+    reject: (reason?: unknown) => void
+  }>()
+  const { startTimer, resetTimer, time } = useTimer(MESSAGE_TIMEOUT_TIME / 1000 - 1)
+  const [validatingProfile, setValidatingProfile] = useState<NodeJS.Timer>()
+  const [isValidated, setIsValidated] = useState<boolean>()
+
+  useEffect(() => {
+    if (signatureResolution) {
```
Comment (andyesp) [inline src/hooks/useForumConnect.ts:37]: `if (signatureResolution && !signState.signing)`

File: src/hooks/useForumConnect.ts [hunk 33]
```diff
@@ -0,0 +1,130 @@
...
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [signatureResolution, signState])
+
+  useEffect(() => {
+    if (time <= 0) {
+      if (validatingProfile) {
+        clearInterval(validatingProfile)
+        setValidatingProfile(undefined)
+      }
+      console.error('Validation Time Expired')
```
Comment (andyesp) [inline src/hooks/useForumConnect.ts:56]: Remove console.error

File: src/hooks/useForumConnect.ts [hunk 34]
```diff
@@ -0,0 +1,130 @@
...
+          resetTimer()
+          signatureResolution.reject(signState.error)
+        }
+      }
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [signatureResolution, signState])
+
+  useEffect(() => {
+    if (time <= 0) {
```
Comment (andyesp) [inline src/hooks/useForumConnect.ts:51]: `if (time <= 0 && validatingProfile)`

File: src/hooks/useForumConnect.ts [hunk 35]
```diff
@@ -0,0 +1,130 @@
...
+        try {
+          const { valid } = await Governance.get().validateProfile()
+          if (valid) {
+            clearInterval(validationChecker)
+            resetTimer()
+            setIsValidated(true)
+            track(SegmentEvent.IdentityCompleted, { address: user, account: 'forum' })
+          }
+        } catch (error) {
+          console.error(error)
```
Comment (andyesp) [inline src/hooks/useForumConnect.ts:104]: Remove console.error

File: src/components/User/UserStats.tsx [hunk 36]
```diff
@@ -26,17 +30,24 @@ const UserAvatar = React.lazy(() => import('./UserAvatar'))
...
       <div className="UserStats__UserInfo">
-        <Mobile>
-          <Username address={address} size="small" className="UserStats__Username" />
-        </Mobile>
-        <NotMobile>
-          <Username address={address} size="medium" className="UserStats__Username" />
-        </NotMobile>
+        <div className="UserStats__UsernameContainer">
+          <Mobile>
+            <Username address={address} size="small" className="UserStats__Username" />
```
[the remark this answers is not in the document]
Comment (ncomerci) [reply src/components/User/UserStats.tsx:43]: I think this is only in profile 🤔

File: src/entities/User/routes.ts (hunk 28 above)
Comment (ncomerci) [reply src/entities/User/routes.ts:39]: It is an endpoint with authentication. If you try to use it without being authenticated, the middleware rejects in advance

image

File: src/entities/User/routes.ts (hunk 29 above)
Comment (ncomerci) [reply src/entities/User/routes.ts:58]: Idem answer

File: src/entities/User/routes.ts (hunk 28 above)
Comment (ncomerci) [reply src/entities/User/routes.ts:39]: And it is the same as we have it everywhere, here is an example:

image

**Review by 1emu — APPROVED** [review]

**Pull request merged**